### DISCUSSION TITLE
gcc source code

### DISCUSSION THREAD

**Discussion by ramiropolla** [SpenceKonde/DxCore#387] [General]
Hi,

Where can I find the source code that was used to generate the builds such as avr-gcc-7.3.0-atmel3.6.1-azduino6-x86_64-pc-linux-gnu.tar.bz2? Stock gcc 7.3.0 doesn't have files such as "device-specs/specs-avr128db28", so this build must have been made with modified sources.

Ramiro

**Comment by WestfW** [comment]
AFAIK, avr-gcc starting in 5.4 or before supports addition of new devices via a "PACKS" mechanism, and the device specs are normally distributed by Microchip separately from the compiler.  http://packs.download.atmel.com/
The normally supported way to use a PACK is via adding -B and -I switches to the compile command, but they can also be added manually to the normal gcc and avr-libc directory structure, as described here: https://www.avrfreaks.net/s/topic/a5C3l000000UaqGEAS/t155412?comment=P-1225272

(I've always assumed that the Arduino and DrAzzy cores have added the packs manually, or perhaps come up with a semi-automatic way of adding them, rather than modifying the core gcc source or build process.  I could be wrong.)

**Comment by MX682X** [comment]
Here you can find the original buildscript: https://github.com/arduino/toolchain-avr
I've build avr-gcc with newer gcc versions recently to try them out, didn't produce any improvements. If you want to read about it, here is the link to the discussion: https://github.com/SpenceKonde/DxCore/discussions/381

**Reply by SpenceKonde** [reply]
I have a slightly modified build script which automatically grabs pack files (though I still need to modify it so it knows which ones to grab), and a whole bunch of text files that I copy/paste console commands and regex find-replace commands. Since I can't cross-compile the binaries, those are still stock 7.3.0 binaries from Arduino7, but the packs get updated when I build my toolchain packages. The process is not automated, for the most part, and is done on a fairly high spec AWS machine that I rent for like the half hour it takes to build all the precompiled shit (standard libraries compiled to .a files) then I take the result and combine it with the Arduino7 versions for each platform, since only the executables for the non-AVR system (which don't change) are problematic for me to create.
